### PR TITLE
bpo-40422: create a common _Py_closerange API

### PR DESCRIPTION
**Pull request by kevans91**
Such an API can be used both for os.closerange and subprocess. For the latter, this yields potential improvement for platforms that have fdwalk but wouldn't have used it there. This will prove even more beneficial later for platforms that have close_range(2), as the new API will prefer that over all else if it's available.

The new API is structured to look more like close_range(2), closing from [start, end] rather than the [low, high) of os.closerange().

https://bugs.python.org/issue40422

Automerge-Triggered-By: @gpshead

### CODE REVIEW

**Comment by the-knights-who-say-ni** [comment]
Hello, and thanks for your contribution!

I'm a bot set up to make sure that the project can legally accept this contribution by verifying everyone involved has signed the [PSF contributor agreement](https://www.python.org/psf/contrib/contrib-form/) (CLA).



## CLA Missing

Our records indicate the following people have not signed the CLA:

@kevans91

For legal reasons we need all the people listed to sign the CLA before we can look at your contribution. Please follow [the steps outlined in the CPython devguide](https://devguide.python.org/pullrequest/#licensing) to rectify this issue.

If you have recently signed the CLA, please wait at least one business day
before our records are updated.


You can [check yourself](https://check-python-cla.herokuapp.com/) to see if the CLA has been received.

Thanks again for the contribution, we look forward to reviewing it!

**Comment by kevans91** [comment]
Anything I can do to help this along?

Also, not sure if this is is necessarily worth adding a news entry, but I can do that if requested.

**Comment by kevans91** [comment]
Ping... same question as last time, anything I can do to help this along?

File: Modules/_posixsubprocess.c [hunk 1]
```diff
@@ -258,22 +258,11 @@ _close_fds_by_brute_force(long start_fd, PyObject *py_fds_to_keep)
...
-        closefrom(start_fd);
-#else
-        for (fd_num = start_fd; fd_num < end_fd; ++fd_num) {
-            /* Ignore errors */
-            (void)close(fd_num);
-        }
-#endif
-    }
+    if (start_fd <= end_fd)
+        _Py_closerange(start_fd, end_fd);
```
Comment (ZackerySpytz) [inline Modules/_posixsubprocess.c]: The [PEP 7](https://www.python.org/dev/peps/pep-0007/#code-lay-out) style guide states that "all new C code requires braces".

File: Modules/posixmodule.c [hunk 2]
```diff
@@ -8696,7 +8711,39 @@ _fdwalk_close_func(void *lohi, int fd)
...
-#endif /* HAVE_FDWALK */
+#endif /* USE_FDWALK */
+
+/* Closes all file descriptors in [first, last], ignoring errors. */
+void
+_Py_closerange(int first, int last)
+{
+#ifdef USE_FDWALK
+    int lohi[2];
+#endif
```
Comment (ZackerySpytz) [inline Modules/posixmodule.c]: This declaration can be moved down, eliminating the need for this extra `#ifdef`.

Comment (kevans91) [reply Modules/posixmodule.c]: Ah, thanks! I missed this part: `This means (amongst many other things) that all declarations must be at the top of a block (not necessarily at the top of function).`

**Review by gpshead — APPROVED** [review]
thanks!

FYI - Linux appears to be gaining a close_range system call this year but writing the code to determine if that syscall is available on the running kernel and use it belongs in a separate PR.
  http://lkml.iu.edu/hypermail/linux/kernel/2008.0/02649.html

**Comment by kevans91** [comment]
> thanks!
> 
> FYI - Linux appears to be gaining a close_range system call this year but writing the code to determine if that syscall is available on the running kernel and use it belongs in a separate PR.
> http://lkml.iu.edu/hypermail/linux/kernel/2008.0/02649.html

Indeed, thanks for taking a look!

I actually have a branch queued up with the rest in the pipeline that was awaiting this review -- I've split it up into configure script change and then preferring it in _Py_closerange: https://github.com/kevans91/cpython/commits/close_range

**Pull request merged**